src/Microsoft.OData.Client/ALinq/ResourceBinder.cs

```diff
@@ -384,19 +387,79 @@ private static void ValidateFilter(Expression exp, string method)
 
                     Debug.Assert(schemaType != null, "SchemaType can not be null.");
 
+                    var keys = schemaType.Key().ToArray();
                     // Obtain the key properties from EdmEntityType, and compare them with keys obtained from predicates
                     // By this time we are sure that keyValuesFromPredicates has unique set (i.e. no duplicates), and has all the keys in it.
                     // So just comparing count with EDMModel's keys is enough for validation
-                    bool allKeysPresent = schemaType.Key().Count() == keyValuesFromPredicates.Keys.Count;
+                    bool allKeysPresent = keys.Length == keyValuesFromPredicates.Keys.Count;
 
                     if (!allKeysPresent)
                     {
                         keyValuesFromPredicates = null;
                         keyPredicates = null;
                     }
+
+                    if (allKeysPresent && keys.Length > 1)
+                    {
+                        // OData Client generates a request that should be the canonical URL, based on the order of keys in the CSDL.
+                        keyPredicates = AdjustKeyExpressionsBasedOnKeyOrders(keys, keyPredicates, expToPropertyNameMap);
+                    }
+                }
+            }
+
+            return keyPredicates;
+        }
+
+        private static List<Expression> AdjustKeyExpressionsBasedOnKeyOrders(IEdmStructuralProperty[] edmKeys,
+            List<Expression> keyPredicates, Dictionary<Expression, string> expToPropertyNameMap)
+        {
+            bool needAdjust = false;
+            for (int i = 0; i < edmKeys.Length; ++i)
+            {
+                IEdmStructuralProperty edmKey = edmKeys[i];
+                Expression keyExpr = keyPredicates[i];
+
+                if (edmKey.Name != expToPropertyNameMap[keyExpr])
+                {
+                    needAdjust = true;
+                    break;
                 }
             }
 
+            if (needAdjust)
+            {
+                List<Expression> orderedKeyPredicates = new List<Expression>();

```

Comment:
Initialize list with known size to avoid potentially resizing list when adding entries:
```suggestion
                List<Expression> orderedKeyPredicates = new List<Expression>(keyPredicates.Count);
```